### PR TITLE
Rewrite CupertinoActionSheet

### PR DESCRIPTION
**Pull request by dkwingsmt**
This PR rewrite `CupertinoActionSheet` to be based on layout widgets instead of the custom layout code in `_CupertinoDialogRenderWidget`, making it much easier to understand. It also fixes a layout deviation from SwiftUI, and add a few tests.

`_CupertinoDialogRenderWidget` is a custom layout widget that contains too much boilerplate code and is hard to understand where is customized. The plan is to replace it with widgets as much as possible. (Since this widget is used by both `CupertinoActionSheet` and `CupertinoAlertDialog`, this PR only removes the branches related to `isActionSheet`).

The need to use custom layout widget probably came from the following difficulties:
1. The parent needs to know whether an action button is being pressed to correctly render or hide each divider, but the action button is provided by the user as a built widget that also needs to respond to the tap gesture.
2. The main sheet's column needs to _prioritize_ allocating the space to the content section before allocating the remaining to the actions section, while also leaving the actions section a minimum height. This is not supported by simply `Column`.
3. The minimum height of the action section is a non-trivial algorithm.

Luckily, I've found out that all these problems are solvable with widgets.

All existing tests pass.

## Pre-launch Checklist

- [ ] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [ ] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [ ] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [ ] I signed the [CLA].
- [ ] I listed at least one issue that this PR fixes in the description above.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [ ] I added new tests to check the change I am making, or this PR is [test-exempt].
- [ ] I followed the [breaking change policy] and added [Data Driven Fixes] where supported.
- [ ] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[Data Driven Fixes]: https://github.com/flutter/flutter/wiki/Data-driven-Fixes

### CODE REVIEW

File: packages/flutter/lib/src/cupertino/dialog.dart [hunk 1]
```diff
@@ -759,49 +755,302 @@ class CupertinoActionSheetAction extends StatelessWidget {
...
+    super.key,
+    this.isCancel = false,
+    required this.onStateChange,
+    required this.child,
   });
 
-  final Widget? child;
+  final bool isCancel;
+
+  /// The callback that is called when the button is tapped.
```
Comment (chunhtai) [inline packages/flutter/lib/src/cupertino/dialog.dart:770]: This comment and variable name don't seem to match

File: packages/flutter/lib/src/cupertino/dialog.dart [hunk 2]
```diff
@@ -759,49 +755,302 @@ class CupertinoActionSheetAction extends StatelessWidget {
   }
 }
 
-class _CupertinoActionSheetCancelButton extends StatefulWidget {
-  const _CupertinoActionSheetCancelButton({
-    this.child,
+// This widget renders the background (pressed or not) of a button, and sends
+// whether the button is pressed to the parent.
```
Comment (chunhtai) [inline packages/flutter/lib/src/cupertino/dialog.dart:759]: a bit confused about this doc string, will this even send when the button is not pressed?

File: packages/flutter/lib/src/cupertino/dialog.dart [hunk 3]
```diff
@@ -759,49 +755,302 @@ class CupertinoActionSheetAction extends StatelessWidget {
...
+class _ActionSheetButtonBackground extends StatefulWidget {
+  const _ActionSheetButtonBackground({
+    super.key,
+    this.isCancel = false,
+    required this.onStateChange,
```
Comment (chunhtai) [inline packages/flutter/lib/src/cupertino/dialog.dart:764]: maybe call this onPressStateChange

File: packages/flutter/lib/src/cupertino/dialog.dart (hunk 2 above)
Comment (chunhtai) [inline packages/flutter/lib/src/cupertino/dialog.dart:759]: maybe
```
// This widget renders the background (pressed or not) of a button and calls `onPressStateChange`
// when user tap down or lift up on the button

File: packages/flutter/lib/src/cupertino/dialog.dart [hunk 4]
```diff
@@ -759,49 +755,302 @@ class CupertinoActionSheetAction extends StatelessWidget {
...
+    );
+  }
+}
+
+// The keys used by the dividers in the action section of an action sheet.
+//
+// The buttons and the dividers are placed alternately, both keyed by indexes.
+// Dividers use this special class so that the two lists of indexes can be told
+// apart.
+class _DividerKey extends ValueKey<int> {
```
Comment (chunhtai) [inline packages/flutter/lib/src/cupertino/dialog.dart:862]: I am curious why this is needed, are you seeing some state lost due if not using this?

File: packages/flutter/lib/src/cupertino/dialog.dart [hunk 5]
```diff
@@ -759,49 +755,302 @@ class CupertinoActionSheetAction extends StatelessWidget {
...
+  @override
+  _ActionSheetMainSheetState createState() => _ActionSheetMainSheetState();
+}
+
+class _ActionSheetMainSheetState extends State<_ActionSheetMainSheet> {
+  int? _pressedIndex;
+  double _topOverscroll = 0;
+  double _bottomOverscroll = 0;
+
+  // Fills the overscroll area at the top and bottom of the sheet. This is
```
Comment (chunhtai) [inline packages/flutter/lib/src/cupertino/dialog.dart:950]: If I understand correctly, this is because some pressed down button will have a different color so this code let's the button paint its background instead of this top level container.

Can we instead paint the basic background in _ActionSheetMainSheet and the pressed button to paint its own color on top?

Comment (chunhtai) [reply packages/flutter/lib/src/cupertino/dialog.dart:950]: also should watch out for one frame delay if you are changing the size based on scroll notification.

File: packages/flutter/test/cupertino/action_sheet_test.dart [hunk 6]
```diff
@@ -575,7 +675,7 @@ void main() {
     await tester.tap(find.text('Go'));
     await tester.pump();
 
-    expect(findScrollableActionsSectionRenderBox(tester).size.height, moreOrLessEquals(112.3));
+    expect(findScrollableActionsSectionRenderBox(tester).size.height, moreOrLessEquals(84.0));
```
Comment (chunhtai) [inline packages/flutter/test/cupertino/action_sheet_test.dart:678]: why this change?

Comment (chunhtai) [reply packages/flutter/test/cupertino/action_sheet_test.dart:678]: same for other in the file

**Review by chunhtai — COMMENTED** [review]
overall I am on board with this refactor, my comments are mostly in the implementation details

File: packages/flutter/test/cupertino/action_sheet_test.dart (hunk 6 above)
Comment (dkwingsmt) [reply packages/flutter/test/cupertino/action_sheet_test.dart:678]: Let me revert this change and propose it in a separate PR.

File: packages/flutter/lib/src/cupertino/dialog.dart (hunk 5 above)
Comment (dkwingsmt) [reply packages/flutter/lib/src/cupertino/dialog.dart:950]: > Can we instead paint the basic background in _ActionSheetMainSheet and the pressed button to paint its own color on top?

I thought about it, but I'm not sure if it will be exactly as good, because both the basic background and the pressed background are translucent. 

> also should watch out for one frame delay if you are changing the size based on scroll notification.

Are you certain there will be a one frame delay, and the delay will cause a overlap and a gap on the overscroll (if that's the concern)?

I don't think that will happen, because to my understanding, after the pointer event, the gesture recognizer update the position of the scroll view with a `setState` and immediately uses the scroll notification to update the overscrolls' sizes. Therefore both change should take place at the same frame, the next frame.

I've also updated the test `Actions section correctly renders overscrolls` to verify the rendering result at the _next frame_ of the gesture, and no overlaps or gaps can be observed.

image

```dart
    await gesture.moveBy(const Offset(0, -300));
    // Test the bottom overscroll. Use `pump` not `pumpAndSettle` to verify the
    // rendering result of the immediate next frame.
    await tester.pump();
    await expectLater(
      find.byType(CupertinoActionSheet),
      matchesGoldenFile('cupertinoActionSheet.overscroll.2.png'),
    );
```

File: packages/flutter/lib/src/cupertino/dialog.dart (hunk 4 above)
Comment (dkwingsmt) [reply packages/flutter/lib/src/cupertino/dialog.dart:862]: I wanted to add keys to each element of the list. Since there are two list of integers, I can't add, for example, `ValueKey<int>(1)` to both the second item and the first divider (I've tested that in this case Flutter will report duplicate key error.)

If you mean that we don't need keys, well, I'm not strong on it either. But I was thinking that adding keys to a list of items might improve the performance, or at least is a good practice. What do you think?

File: packages/flutter/lib/src/cupertino/dialog.dart (hunk 1 above)
Comment (dkwingsmt) [reply packages/flutter/lib/src/cupertino/dialog.dart:770]: I've changed it to:
```dart
  /// Called when the user taps down or lifts up on the button.
  ///
  /// The boolean value is true if the user is tapping down on the button.
```

File: packages/flutter/lib/src/cupertino/dialog.dart (hunk 2 above)
Comment (dkwingsmt) [reply packages/flutter/lib/src/cupertino/dialog.dart:759]: Thanks, I've changed it to
```dart
// Renders the background of a button (both the pressed background and the idle
// background) and reports its state to the parent with `onPressStateChange`.
```

(The exact definition of "the state" is explained in `onPressStateChange`.) What do you think of it?

File: packages/flutter/lib/src/cupertino/dialog.dart (hunk 5 above)
Comment (chunhtai) [reply packages/flutter/lib/src/cupertino/dialog.dart:950]: > Are you certain there will be a one frame delay, and the delay will cause a overlap and a gap on the overscroll (if that's the concern)?

So I looks more into the overscroll notification code, it looks like the call loop will call into to the update method when as soon as user drag. so I think there shouldn't be a delay. sorry for the confusion.

> I thought about it, but I'm not sure if it will be exactly as good, because both the basic background and the pressed background are translucent. We'll have to simulate one translucent color by combining two translucent colors.

My original thought was trying to remove this weird additional padding, needed when overscroll. but I do agree it may be hard to find the right color to paint on top if they are translucent. The only potentional problem I can think of now since there won't be a frame delay will then be possible anti-aliasing line in the gap. but I don't think that is something we need to worry about not. 

LGTM for this approach then

File: packages/flutter/lib/src/cupertino/dialog.dart (hunk 4 above)
Comment (chunhtai) [reply packages/flutter/lib/src/cupertino/dialog.dart:864]: key in a multichildrenderobject will only matter if you reorder or add or remove child in between update and you want to maintain state, otherwise it is the same as unkeyed, so I will probably not bother adding one for the divider

Comment (dkwingsmt) [reply packages/flutter/lib/src/cupertino/dialog.dart]: Done, removed!

File: packages/flutter/test/cupertino/action_sheet_test.dart (hunk 6 above)
Comment (dkwingsmt) [reply packages/flutter/test/cupertino/action_sheet_test.dart]: I've reverted these changes (and changed the minimum height from 84.0 to 84.3). Now all previous tests pass without any adjustment!

**Comment by dkwingsmt** [comment]
I've addresses all comments and made changes so that all existing tests pass without adjustment.

**Review by chunhtai — APPROVED** [review]
LGTM
